[Spring Core] 서가은 미션 제출합니다.

gaxunsxo opened the pull request:
안녕하세요~ 리뷰 잘 부탁드립니다 :)
이번 core 미션은 먼저 예약 기능과 비슷하게 시간 모델을 만들어 respository, service, controller 를 구현하였습니다.
이후엔 테이블 스키마를 다시 정의하고, String으로 되어 있던 시간을 Time 타입으로 변경하였습니다.
또 학습한 계층화 이론을 토대로 service에 있던 예약 관리 기능을 repository와 service로 분리하여 리팩토링을 진행하였습니다.
처음 다루어 보는 개념이 많았는데, 리뷰어님이 미션 진행하시면서 알게 된 개념이 있다면 공유해보면 좋을 것 같습니다!

gaxunsxo committed:
feat: 데이터베이스 스키마에 time 테이블 생성

gaxunsxo committed:
feat: Time 모델 클래스 정의

gaxunsxo committed:
feat: CRUD 작업을 위한 TimeRepository 클래스 구현

gaxunsxo committed:
feat: 비즈니스 로직을 위한 TimeService 클래스 구현

gaxunsxo committed:
feat: API 엔드포인트를 위한 TimeController 클래스 생성

gaxunsxo committed:
test: 8단계 - 시간 관리 기능

gaxunsxo committed:
fix: 테이블 스키마 재정의

gaxunsxo committed:
fix: 시간 타입을 Time 객체로 수정

gaxunsxo committed:
feat: 계층화 리팩터링

gaxunsxo committed:
test: 8/9단계 - 기존 코드 수정 및 계층화 리팩터링

kwonyj1022 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -26,7 +21,7 @@ public ReservationController(ReservationService reservationService) {
+        return "new-reservation";
     }
 
     @PostMapping("/reservations")
```
예약 추가 api가 동작하지 않아요! 🥲

gaxunsxo replied:
JSON 요청을 Reservation 객체로 변환하는 과정에서 Time 객체를 생성할 수 없는 에러가 발생해서, Time 클래스에 id를 인자로 받는 생성자를 추가하고 Reservation 객체와 JSON 요청을 매핑하였습니다!

kwonyj1022 commented on src/main/java/roomescape/controller/TimeController.java:
```diff
@@ -0,0 +1,57 @@
+    @Autowired
+    public TimeController(TimeService timeService) {
+        this.timeService = timeService;
+    }
```
```suggestion
    public TimeController(TimeService timeService) {
        this.timeService = timeService;
    }
```
생성자 주입의 경우 `@Autowired` 어노테이션을 사용하지 않아도 됩니다!

kwonyj1022 commented on src/main/java/roomescape/controller/TimeController.java:
```diff
@@ -0,0 +1,57 @@
+    @DeleteMapping("/times/{id}")
+    @ResponseBody
+    public ResponseEntity<?> deleteTime(@PathVariable Long id) {
+        boolean isDeleted = timeService.deleteTime(id);
+        if (isDeleted) {
+            return ResponseEntity.noContent().build();
+        } else {
+            return ResponseEntity.badRequest().build();
+        }
```
```suggestion
        boolean isDeleted = timeService.deleteTime(id);
        if (isDeleted) {
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.badRequest().build();
```
else문 대신 early return 방식을 사용하는 것이 더 좋습니다!


또, isDeleted가 false인 상황이 예외적인 상황이기 때문에 이를 강조하기 위해 아래와 같이 isDeleted가 false인지를 if문 넣어도 좋을 것 같아요!
```suggestion
        boolean isDeleted = timeService.deleteTime(id);
        if (!isDeleted) {
            return ResponseEntity.badRequest().build();
        } 
        return ResponseEntity.noContent().build();
```


또는 앞서 말했듯이 '예외'인 상황이기 때문에 서비스 레이어나 컨트롤러 레이어 중 적합한 위치에서 예외를 발생시키고 이 exception을 GlobalExceptionHandler에서 핸들링해주는 방식도 좋을 것 같습니다!

gaxunsxo replied:
네! if문 조건을 수정하고, early return으로 수정하였습니다. 
예외로 처리하는 방법도 참고하겠습니다!

kwonyj1022 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,64 +1,38 @@
     public boolean deleteReservation(Long id) {
-        String sql = "DELETE FROM reservation WHERE id = ?";
-        int rowsAffected = jdbcTemplate.update(sql, id);
+        int rowsAffected = reservationRepository.deleteById(id);
         return rowsAffected > 0;
```
Time의 경우 delete 후 영향받은 rows 개수에 따른 boolean 반환을 repository에서 수행하는 반면, Reservation은 service에서 수행하고 있어요. 가은님은 해당 역할이 repository와 service 중 어느 계층에 적합하다고 생각하시나요?

gaxunsxo replied:
삭제 로직을 service 계층에서 수행하니까, time과 reservation 모두 service에서 동일한 역할을 하도록 수정하는 게 더 낫겠네요. 반영하겠습니다!

kwonyj1022 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,64 +1,38 @@
 import java.util.List;
 
 @Service
 public class ReservationService {
-    private final JdbcTemplate jdbcTemplate;
+
+    private final ReservationRepository reservationRepository;
```
서비스 계층을 사용해주셨네요! service 계층의 역할은 무엇이라고 생각하시나요?

gaxunsxo replied:
비즈니스 로직을 처리하고, repository 계층과 controller 계층 사이의 중간 역할을 수행한다고 생각합니다.

kwonyj1022 commented on src/main/java/roomescape/controller/TimeController.java:
```diff
@@ -0,0 +1,57 @@
+    @PostMapping("/times")
+    @ResponseBody
+    public ResponseEntity<Time> createTime(@RequestBody Time time) {
+        if(!timeService.isValidTime(time)) {
+            return ResponseEntity.badRequest().build();
+        }
+
+        Time newTime = timeService.createTime(time);
```
controller에서도 isValidTime() 메서드를 사용하고, service의 createTime() 메서드 내부에서도 isValidTime() 메서드를 사용하고 있는 중복이 있네요! isValidTime()은 컨트롤러의 역할인가요? 아니면 서비스의 역할인가요?

gaxunsxo replied:
데이터의 유효성을 검증하는 건 service 의 역할이니까, 각 계층의 역할에 맞게 controller 내에서 isValidTime() 메서드를 사용하지 않는게 좋을 것 같습니다!

kwonyj1022 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,64 +1,38 @@
-import java.sql.PreparedStatement;
 import java.util.List;
 
 @Service
 public class ReservationService {
```
`@Transactional` 어노테이션에 대해 학습해보시기를 추천드려요!

kwonyj1022 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -154,4 +159,58 @@ public class MissionStepTest {
         assertThat(countAfterDelete).isEqualTo(0);
     }
+
+    @Test
```
전체 테스트를 돌렸을 때 실패하는 테스트가 있어요. 요구사항이 변경되면서 실패하게 되는 테스트인데, 저는 개인적으로 모든 테스트가 통과하는 pr이 머지되는 것이 좋다고 생각하기 때문에 `@Disabled` 어노테이션을 이용해서 테스트 수행에서 제외시키는 방법을 사용하곤 합니다. 가은님은 실패하는 테스트가 있는 pr 머지에 대해 어떻게 생각하시나요?

gaxunsxo replied:
실패하는 테스트가 있는 pr이라면 머지하는 사람의 입장에선 문제가 있는 코드라고 생각될 수 있을 것 같습니다.  @ `@Disabled` 어노테이션은 몰랐는데, 이번 기회에 알게 되었습니다!

kwonyj1022 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -39,7 +34,7 @@ public ResponseEntity<Reservation> createReservation(@RequestBody Reservation re
+               .body(newReservation);
     }
 
     @DeleteMapping("/reservations/{id}")
```
가은님 코드에 `NotFoundReservationException`이 있는데 이 예외는 어디에서도 사용되지 않고 있어요. 해당 에외를 이용해 보는 것은 어떨까요?

kwonyj1022 reviewed (changes requested):
마지막 미션 수고 많으셨어요!
예약 추가 api가 동작하지 않는 것 같은데 확인 부탁드려요!

gaxunsxo committed:
fix: 예약 추가 API 오류 해결

gaxunsxo committed:
fix: 불필요한 어노테이션 삭제

gaxunsxo committed:
fix: 시간 삭제 시, early return 방식으로 수정

gaxunsxo committed:
refactor: 서비스 계층으로 로직 이동 및 레포지토리 메서드 수정
- TimeRepository의 deleteById 메서드를 수정하여 삭제된 행의 수를 반환하도록 변경
- TimeService에서 deleteById의 결과를 처리하고 성공 여부를 boolean으로 반환하도록 업데이트
- 삭제 작업 처리와 리포지토리 및 서비스 계층 간의 책임 분리 일관성 유지

gaxunsxo committed:
refactor: Time 검증 로직을 서비스 계층으로 이동

gaxunsxo committed:
fix: 필요하지 않은 테스트 제외

gaxunsxo committed:
fix: 유효하지 않은 예약에 대해 NotFoundReservationException 던지기

gaxunsxo committed:
fix: 필요하지 않은 선언 제거
